dmipy/utils/invariants.py

```diff
@@ -0,0 +1,576 @@
+                   [2, 6, 6, 8],
+                   [2, 6, 8, 8],
+                   [2, 8, 8, 8],
+                   [4, 4, 4, 4],
+                   [4, 4, 4, 6]]
+    return l_list
+
+
+def get_invariants(sh_coef, l_list, sh_base='mrtrix', normalize = True):
+    r""" Returns the rotation invariants features specified in l_list for the
+    given SH coefficients
+
+    Parameters
+    ----------
+    sh_coef : array,
+        array containing the SH coefficients. It can be multidimensional but
+        the last dimension must contain the SH coefficients. E.g. a three
+        dimensional image with dimension x, y, and z the shape of sh_coef
+        must be [x,y,z,n_c] with n_c the number of SH coefficients.
+    l_list : list,
+        the list of lists of desired invariants
+    sh_base : string,
+        the base in which the sh_coef are represented. Admitted values are
+        'mrtrix', 'dipy', and 'gaunt'
+    normalize : boolean,
+        True if the invariants have to be normalized between [-1,1]
+
+    Returns
+    -------
+    invariants : array
+        array containing the desired invariants
+    """
+    if not(sh_base in ['mrtrix', 'dipy']):
+        raise ValueError(
+            'sh_base must be either "mrtrix" or "dipy"')
+
+
+    l_max = np.max([np.max(l) for l in l_list])
+    if l_max > 8:
+        raise ValueError(
+            'maximum degree in l_list is greater than 8')
+
+    len_max = np.max([len(l) for l in l_list])
+    if len_max > 5:
+        raise ValueError(
+            'maximum power in l_list is greater than 5')
+
+    odd_number = np.sum([np.sum(np.array(l)%2) for l in l_list])
+    if odd_number > 0:
+        raise ValueError(
+            'all of the degrees in l_list must be even')
+
+    N = sh_coef.shape[-1]
+    sh_order = int((-3 + np.sqrt(1+8*N))/2.0)
+
+    if sh_order < l_max:
+        raise ValueError(
+            'SH order of the coefficients is lower than the degree of the required invariants')
+
+    RG = load_npz(join(DATA_PATH, 'real_gaunt_coefficients.npz')).toarray()[0,:]
+    coef = convert_sh(sh_coef, in_base=sh_base)
+
+    inv_number = len(l_list)
+
+    if len(coef.shape) > 1:
+        shape = list(coef.shape[:-1])+[inv_number]
+    else:
+        shape = inv_number
+
+    invariants = np.zeros(shape)
+
+    s4p = np.sqrt(4*np.pi)
+    is4p = 1.0/np.sqrt(4*np.pi)
+    MTV = []
+    for i, l_i in enumerate(l_list):
+        len_l = len(l_i)
+        if len_l==1:
+            if l_i[0] == 0:
+                mtv = 1
+                invariants[..., i] = coef[...,0]*s4p
+
+
+        if len_l==2:
+            if l_i[0] == l_i[1]:
+                ll = l_i[0]
+                mtv = (2*ll+1)/(4*np.pi)
+                start = (ll-1)*(ll)//2
+                stop = (ll+1)*(ll+2)//2
+                invariants[...,i] = np.sum(coef[..., start:stop]**2, -1)
+
+        if len_l==3:
+            l1 = l_i[0]
+            l2 = l_i[1]
+            l3 = l_i[2]
+            mtv = 0
```

Comment:
I think you can create `mtv` just once at the beginning of the for loop and not do it for every case separately.

Reply:
Maybe it will be a bit faster. I have to try it...

Reply:
I don't think the speed increase will be significant, just fewer lines ;-)